pkg/runner/step_context.go

```diff
@@ -410,150 +407,170 @@ func (sc *StepContext) runAction(actionDir string, actionPath string) common.Exe
-			return rc.execJobContainer(containerArgs, sc.Env)(ctx)
+			return sc.execAsNode(ctx, step, actionDir, rc, containerActionDir, actionName, actionPath, action)
 		case model.ActionRunsUsingDocker:
-			var prepImage common.Executor
-			var image string
-			if strings.HasPrefix(action.Runs.Image, "docker://") {
-				image = strings.TrimPrefix(action.Runs.Image, "docker://")
-			} else {
-				image = fmt.Sprintf("%s:%s", regexp.MustCompile("[^a-zA-Z0-9]").ReplaceAllString(actionName, "-"), "latest")
-				image = fmt.Sprintf("act-%s", strings.TrimLeft(image, "-"))
-				image = strings.ToLower(image)
-				contextDir := filepath.Join(actionDir, actionPath, action.Runs.Main)
-
-				anyArchExists, err := container.ImageExistsLocally(ctx, image, "any")
-				if err != nil {
-					return err
-				}
-
-				correctArchExists, err := container.ImageExistsLocally(ctx, image, rc.Config.ContainerArchitecture)
-				if err != nil {
-					return err
-				}
+			return sc.execAsDocker(ctx, action, actionName, actionDir, actionPath, rc, step)
+		case model.ActionRunsUsingComposite:
+			return sc.execAsComposite(ctx, step, actionDir, rc, containerActionDir, actionName, actionPath, action)
+		default:
+			return fmt.Errorf(fmt.Sprintf("The runs.using key must be one of: %v, got %s", []string{
+				model.ActionRunsUsingDocker,
+				model.ActionRunsUsingNode12,
+				model.ActionRunsUsingComposite,
+			}, action.Runs.Using))
+		}
+	}
+}
 
-				if anyArchExists && !correctArchExists {
-					wasRemoved, err := container.RemoveImage(ctx, image, true, true)
-					if err != nil {
-						return err
-					}
-					if !wasRemoved {
-						return fmt.Errorf("failed to remove image '%s'", image)
-					}
-				}
+func (sc *StepContext) execAsDocker(ctx context.Context, action *model.Action, actionName string, actionDir string, actionPath string, rc *RunContext, step *model.Step) error {
+	var prepImage common.Executor
+	var image string
+	if strings.HasPrefix(action.Runs.Image, "docker://") {
+		image = strings.TrimPrefix(action.Runs.Image, "docker://")
+	} else {
+		image = fmt.Sprintf("%s:%s", regexp.MustCompile("[^a-zA-Z0-9]").ReplaceAllString(actionName, "-"), "latest")
+		image = fmt.Sprintf("act-%s", strings.TrimLeft(image, "-"))
+		image = strings.ToLower(image)
+		contextDir := filepath.Join(actionDir, actionPath, action.Runs.Main)
 
-				if !correctArchExists {
-					log.Debugf("image '%s' for architecture '%s' will be built from context '%s", image, rc.Config.ContainerArchitecture, contextDir)
-					prepImage = container.NewDockerBuildExecutor(container.NewDockerBuildExecutorInput{
-						ContextDir: contextDir,
-						ImageTag:   image,
-						Platform:   rc.Config.ContainerArchitecture,
-					})
-				} else {
-					log.Debugf("image '%s' for architecture '%s' already exists", image, rc.Config.ContainerArchitecture)
-				}
-			}
+		exists, err := container.ImageExistsLocally(ctx, image, "any")
+		if err != nil {
+			return err
```

Comment:
This also seems like a mistaken regress of #619

Reply:
Yes, sorry about these.
I hope I've fixed them in https://github.com/nektos/act/pull/566/commits/a6fa584e71dd0bcbef366605e03de3e9cd3fbcbe